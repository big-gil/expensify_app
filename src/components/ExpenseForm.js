import React from "react";
import moment from "moment";
// import { SingleDatePicker } from "react-dates";
import { connect } from "react-redux";

const now = moment().valueOf();

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      note: props.expense ? props.expense.note : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      createDate: props.expense ? props.expense.createDate : now,
      error: ""
    };
  }

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = e => {
    const createDate = e.target.value;
    this.setState(() => ({ createDate }));
  };

  onSubmitForm = e => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please enter a description and an amount."
      }));
    } else {
      this.setState(() => ({ error: "" }));
      this.props.onSubmit({
        description: this.state.description,
        amount: Math.round(parseFloat(this.state.amount, 10) * 100),
        note: this.state.note,
        createDate: this.state.createDate
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmitForm}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <input
            type="number"
            placeholder="Date"
            value={this.state.createDate}
            onChange={this.onDateChange}
          />
          <textarea
            value={this.state.note}
            placeholder="Add a note for your expense (optional)"
            onChange={this.onNoteChange}
          />
          <input type="submit" value="Add Expense" />
        </form>
      </div>
    );
  }
}
