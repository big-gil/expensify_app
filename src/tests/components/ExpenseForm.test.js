import React from "react";
import { shallow } from "enzyme";
import ExpenseForm from "../../components/ExpenseForm";
import { expenses } from "../fixtures/expenses";

test("Expense Form Test 1 - Expense Form Render", () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test("Expense Form Test 2 - Expense Form with Data", () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test("Expense Form Test 3 - Invalid Form submision", () => {
  const expense = { description: "", amount: 0 };
  const wrapper = shallow(<ExpenseForm expense={expense} />);
  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });
  expect(wrapper.state("error").length).toBeGreaterThanOrEqual(0);
  expect(wrapper).toMatchSnapshot();
});

test("Expense Form Test 4 - Description Changed", () => {
  const value = "new description";
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("input").at(0).simulate("change", {
    target: { value }
  });
  expect(wrapper.state("description")).toBe(value);
});

test("Expense Form Test 5 - Note Changed", () => {
  const value = "A simple note.";
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("textarea").simulate("change", {
    target: { value }
  });
  expect(wrapper.state("note")).toBe(value);
});

test("Expense Form Test 6 - Amount Changed", () => {
  const value = "25.40";
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("input").at(1).simulate("change", {
    target: { value }
  });
  expect(wrapper.state("amount")).toBe(value);
});

test("Expense Form Test 7 - Amount Changed Invalid", () => {
  const value = "25.5540";
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find("input").at(1).simulate("change", {
    target: { value }
  });
  expect(wrapper.state("amount")).toBe("");
});

test("Expense Form Test 8 - On submit prop with valid data", () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />
  );

  wrapper.find("form").simulate("submit", {
    preventDefault: () => {}
  });

  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: expenses[0].descripton,
    amount: expenses[0].amount,
    note: expenses[0].note,
    createDate: expenses[0].createDate
  });
});
