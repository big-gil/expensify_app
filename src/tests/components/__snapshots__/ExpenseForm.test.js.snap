// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expense Form Test 1 - Expense Form Render 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Date"
      type="number"
      value={0}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value=""
    />
    <input
      type="submit"
      value="Add Expense"
    />
  </form>
</div>
`;

exports[`Expense Form Test 2 - Expense Form with Data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value="5000"
    />
    <input
      onChange={[Function]}
      placeholder="Date"
      type="number"
      value={-259200000}
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
      value="note2"
    />
    <input
      type="submit"
      value="Add Expense"
    />
  </form>
</div>
`;

exports[`Expense Form Test 3 - Invalid Form submision 1`] = `
<div>
  <p>
    Please enter a description and an amount.
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Description"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Amount"
      type="text"
      value="0"
    />
    <input
      onChange={[Function]}
      placeholder="Date"
      type="number"
    />
    <textarea
      onChange={[Function]}
      placeholder="Add a note for your expense (optional)"
    />
    <input
      type="submit"
      value="Add Expense"
    />
  </form>
</div>
`;
